### ISSUE TITLE
Improve DDEV's packaging of gsudo

### ISSUE THREAD

**Issue by gerardog** [ddev/ddev#5324]
Discussion with @gerardog regarding this in:
* https://github.com/gerardog/gsudo/pull/299
* https://github.com/gerardog/gsudo/issues/296

_Originally posted by @rfay in https://github.com/ddev/ddev/issues/5314#issuecomment-1708813273_
            
---
Hi @rfay, I wanted to understand better how ddev uses gsudo, so I cloned every ddev repo:

```
$user = "ddev"
$apiUrl = "https://api.github.com/users/$user/repos?per_page=100"
$outputDir = "C:\git\ddev"

$response = Invoke-RestMethod -Uri $apiUrl
foreach ($repo in $response) {
    $repoName = $repo.name
    $repoUrl = $repo.clone_url
    Write-Host "Cloning $repoName from $repoUrl"
    git clone $repoUrl "$outputDir\$repoName"
}
```
and search for the gsudo literal... I mostly found it's used in:

- docs (*.md)
- install scripts
- choco package as a dependency
- nullsoft installer

So, I didn't find a place where gsudo is used as part of the actual application... Am I right??
If that is the case, is it that you just want the end user to have gsudo available for convenience?

I guess I should first understand if gsudo is a hard dependency, a must-have, or a desired convenience tool. The difference is the first one requires gsudo, the later can prompt the user if wants to install it.

The friendly way of installing gsudo is thru the `.msi` file... It has the convenience that the user gets the complete experience, 
 plus a clean uninstall, can upgrade anytime later using winget upgrade. (Yeah, winget upgrades even if the msi is installed manually), etc...

Currently gsudo is portable but may misbehave if two different versions are running at the same time. That is because as one protection against hackers is that an elevated gsudo will only accept calls from the same `gsudo.exe` file. Running multiple versions simultaneously may fail, hence is prefered to have only one gsudo install. This problem may be mitigated with a code change, but currently is not.

We can dig into the alternatives which I think are:
- Download the `.msi` and install gsudo from ddev installer...
- Make ddev launch gsudo installer... (either .msi or the .ps1 [install script](github.com/gerardog/gsudo/tree/master/installgsudo.ps1) which has an possible inconvenient 'press a key' wait on the end that we can remove.
- Opting for the portable way, by a) mitigating the problem mentioned above and b) embedding a non-virus flagged version inside the installer.

**Comment by rfay** [comment]
Thanks for the response here! Yes, gsudo is used as part of the application when escalation is required to add to the hosts file in WSL2 or traditional Windows. It's used as `sudo` though. 

https://github.com/ddev/ddev/blob/16870030d464e7bb64a569685b65871ce5669469/pkg/ddevapp/hostname_mgt.go#L39-L45

https://github.com/ddev/ddev/blob/16870030d464e7bb64a569685b65871ce5669469/pkg/ddevapp/hostname_mgt.go#L241-L244

**Comment by gerardog** [comment]
Hey, that was a fast response, check out I edited the OP...
Ohh so you are calling it `sudo.exe`?, yeap that works too, the installers create aliases for `sudo`... 

If you are fine to embed a portable version, I suggest you may:
- wait for the fix of the multiple simultaneous gsudo versions
- upload a specific `gsudo.exe` to the ddev installers repo, maybe named as `sudo.exe`, for example into `/third_party/gsudo`
   - do not pickup the latest and greatest version, pick the previous so it already has some popularity in online AV vendors
   - check it has a score of 0 in [virus total](https://www.virustotal.com/gui/file/cf98ddac2fba4874ea695fd607475afd7cade8d4f69cf2023eba2a2197eec8a6/detection/f-cf98ddac2fba4874ea695fd607475afd7cade8d4f69cf2023eba2a2197eec8a6-1693175026)
   - you dont need to update gsudo every day, you can do every 6 months or so.

**Comment by rfay** [comment]
Thanks so much for this. Yeah, most of the logic around hosts file manipulation is generic for linux/mac/windows. The only thing that's different is knowing whether sudo is available (on Windows). The tricky thing that happens is that WSL2 users need to update the *windows* side hosts file for best experience, so `sudo.exe` is used there in WSL2 to do that.

**Comment by rfay** [comment]
Another option for WSL2 is a bit simpler, outside scope of this probably:
* Use gsudo.exe directly and require it to be installed.
* Use an explicit hosts file editor on the Windows side (`sudo ddevfixhosts`) or something to edit the windows-side hosts file instead of using `ddev.exe`, which is way overkill for this simple usage.

**Comment by gerardog** [comment]
> The tricky thing that happens is that WSL2 users need to update the _windows_ side hosts file for best experience, so `sudo.exe` is used there in WSL2 to do that.

Technically you don't need `gsudo` for that... you can do: `powershell.exe -c "Start-Process -Verb runas -FilePath notepad -ArgumentList 'C:\Windows\System32\drivers\etc\hosts'"` from bash:

![image](https://github.com/ddev/ddev/assets/3901474/136aed97-0df4-447a-889a-46d9579f2e7b)

It shows a uac popup and then notepad opens elevated.

Looking at [hostname_mgt.go](https://github.com/ddev/ddev/blob/16870030d464e7bb64a569685b65871ce5669469/pkg/ddevapp/hostname_mgt.go#L39-L45) I think you may create a 20 lines powershell script that does the `ddevfixhosts` part.

**Comment by rfay** [comment]
Wow, thanks for that! So notepad just knows to UAC when accessing a privileged file? But how do you actually do an edit automatically in that case?It seems this would just bring up notepad and then people would have to know what to do.

**Comment by gerardog** [comment]
Lol No! Notepad is terribly dumb! The trick is to start a process (call CreateProcess API) with parameter `Verb=RunAs` which is the official windows way to elevate the process. Here I am using Powershell.exe (built-in on every windows 10 & 11) to call CreateProcess (via the `start-process` powershell cmdlet) so I can make it into one line.

Tearing apart that line: 

```  powershell
powershell.exe        # Calling something ending in .exe will invoke in the windows world (escape wsl linux)
       -c                     # here comes a powershell command
       "Start-Process # documented here https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/start-process?view=powershell-7.3
      -Verb runas           # show the UAC Popup!
      -FilePath notepad # Run notepad or whatever windows EXE file
      -ArgumentList 'C:\Windows\System32\drivers\etc\hosts'" # passing this arguments
```

**Comment by rfay** [comment]
Thanks. So at this point, we'd just run `PowerShell.exe -Verb runas -FilePath ddev.exe -ArgumentList 'hostname', `domain-name-to-add.ddev.site` or something like that.

**Comment by gerardog** [comment]
What if we create a `ddevfixhosts.ps1` powershell file? 
- ddev shall just run `powershell.exe -File ddevfixhosts.ps1 /some/file/with/current/status` ( 

check this out

![image](https://github.com/ddev/ddev/assets/3901474/3eb0fd5d-2cdd-4a86-9e93-d8b3e9d9248a)

**Comment by gerardog** [comment]
>  So at this point, we'd just run PowerShell.exe -Verb runas -FilePath ddev.exe -ArgumentList 'hostname', domain-name-to-add.ddev.site` or something like that.

yes that could work to, instead of a powershell script, make `ddev.exe` fix the things

**Comment by rfay** [comment]
Related:
* https://github.com/ddev/ddev/issues/6663

If we figured out a way to do name resolution (on the windows side) without modifying /etc/hosts, we wouldn't even need escalation any more. 

* https://github.com/ddev/ddev/issues/6440

Probably same as here.
